### PR TITLE
Classifier: preserve pending classifications

### LINKED ISSUE #3016: Repeat subjects on workflows with sequential classification
## Volunteers on sequential-style workflows are seeing repeat subjects
Note that this is a recurrence of an old issue we thought was fixed. Possibility that this is related to Done & Talk (see Talk comments linked below).

## To Reproduce
Have not yet been able to reproduce

Applicable Panoptes resource IDs (project, workflow, etc) to demonstrate the issue:
HMS NHS `11429`
- workflow `18613` & `18611`
Davy Notebooks Project `9006`
- workflow `21009`

## Expected behavior
Subjects should not be delivered again once classified. If they do, they should have an Already Seen or Finished banner.

## Additional context
April 1 (HMS NHS): https://www.zooniverse.org/projects/msalmon/hms-nhs-the-nautical-health-service/talk/3678/2393089?comment=3940218 -- update from April 11, user says it's happened about 3 times since initially reporting this on April 1

April 1 (HMS NHS): https://www.zooniverse.org/projects/msalmon/hms-nhs-the-nautical-health-service/talk/3678/2398525 reporter says repeats are happening at the start of a session, happening semi-frequently, always when choosing Done & Talk:
> When I finish a session (Windows 10, internet explorer, desktop), I click on Done/Talk and close the site before moving on. In the last 4 weeks or so when I start a new session, I have been given the last entered page (of the previous session) about 14 times now. If I refresh or shut down the site, the same page comes up when I start again.
However, I no longer click on the Done/Talk option at the end of a session, clicking the Done option instead and then shut down. That seems to have stopped the issue.

April 11 (Davy): https://www.zooniverse.org/projects/humphrydavy/davy-notebooks-project/talk/2586/2084668?comment=3958495 -- reporter says later in the day the behavior stopped, but the end of the thread has another report from April 11 of the same behavior.

### LINKED ISSUE #4725: Maria Edgeworth Letters: reports of missing classifications
## Package
lib-classifier

## Describe the bug
Volunteer `hoskinml` reports transcribing subject [85843380](https://www.zooniverse.org/projects/mariaedgeworthletters/maria-edgeworth-letters/talk/subjects/85843380) twice: once on 12th May and again on 13th May. We have a classification from them on 13th May (classification 489655460) but not 12th May.

They've also transcribed subject [87415893](https://www.zooniverse.org/projects/mariaedgeworthletters/maria-edgeworth-letters/talk/subjects/87415893) twice, most recently on 17th May, but neither classification shows up as linked to their account. The classification history for that subject doesn’t have a classification on 17th May either. 

They’ve also reported, on Talk, that they didn’t see their daily classification count increase after pressing Done. 

## Expected behavior
Volunteer classifications should be linked to their account, when they're logged in.

Failed classifications should be queued in local browser storage and retried. Classifications should never be lost. 

When you press either Done or Done & Talk, this callback should run and increment your classification count in the browser. 
https://github.com/zooniverse/front-end-monorepo/blob/a356fedf4b26c7c88cc5b9c45f97c8509a89ae15/packages/app-project/src/screens/ClassifyPage/components/ClassifierWrapper/ClassifierWrapper.js#L44-L54

## Additional context

We've previously seen this bug in #2574, where it turned out that classifications from logged-in user accounts were submitted with null or empty authorisation headers.

https://github.com/zooniverse/front-end-monorepo/issues/3016#issuecomment-1098849554 describes a bug where Done & Talk can interrupt the pending classification queue, deleting any unsaved classifications.

### PR DESCRIPTION
**Pull request by eatyourgreens**
- Rewrite the `ClassificationQueue` tests to mock the Panoptes API with 'nock'.
- Add tags to Sentry logs.
- Use `classification.id` to uniquely identify classifications.
- Fix a broken map in `classificationQueue.flushToBackend()`.
- Remove successful classifications from the queue, rather than deleting the whole queue then rebuilding it.

_Please request review from `@zooniverse/frontend` team or an individual member of that team._ 

## Package
lib-classifier

## Linked Issue and/or Talk Post
- may close #4725.
- may close #3016.

## How to Review
The easiest way to check this in a browser is probably with the dev classifier. Verify (in the Network tab) that Done & Talk aborts your classification before it can be saved, but that the classification is kept in local storage, to be retried the next time that you load the classifier or press Done.
https://local.zooniverse.org:8080/?project=908&workflow=3223

# Checklist
_PR Creator - Please cater the checklist to fit the review needed for your code changes._
_PR Reviewer - Use the checklist during your review. Each point should be checkmarked or discussed before PR approval._

## General
- [x] Tests are passing locally and on Github
- [ ] Documentation is up to date and changelog has been updated if appropriate
- [x] You can `yarn panic && yarn bootstrap` or `docker-compose up --build` and FEM works as expected
- [ ] FEM works in all major desktop browsers: Firefox, Chrome, Edge, Safari (Use Browserstack account as needed)
- [ ] FEM works in a mobile browser

## General UX
Example Staging Project: [i-fancy-cats](https://local.zooniverse.org:3000/projects/brooke/i-fancy-cats)
- [x] All pages of a FEM project load: Home Page, Classify Page, and About Pages
- [x] Can submit a classification
- [ ] Can sign-in and sign-out
- [x] The component is accessible
  - Can be used with a screen reader [BBC guide to VoiceOver](https://bbc.github.io/accessibility-news-and-you/assistive-technology/testing-steps/voiceover-mac.html)
  - Can be used from the keyboard [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - It is passing accessibility checks in the storybook

## Bug Fix
- [x] The PR creator has listed user actions to use when testing if bug is fixed
- [x] The bug is fixed
- [x] Unit tests are added or updated

## Refactoring
- [x] The PR creator has described the reason for refactoring
- [ ] The refactored component(s) continue to work as expected

### CODE REVIEW

**Review by goplayoutside3 — COMMENTED** [review]
To test this PR, I ran dev-classifier locally and loaded the project linked in the PR description. I answered the task and hit Done & Talk. Now `localStorage` has one failed classification in its array because /talk cannot load when running the classifier locally. I hit the browser back button to re-load the classifier, a different subject loaded, I completed the classification and this time hit “Done”. Now the failed classifications array in `localStorage` is now empty as expected ✅ 

I can replicate the same failed classifications process when running app-project with i-fancy-cats, but I have one concern about the failed classifications array. When viewing i-fancy-cats, let's say I navigate away by clicking Done & Talk (trying to visit /talk will error when running FEM locally). I click the browser back button, but now I’m served the same subject. From here I have to complete a second classification on the same subject to try emptying the failed classifications array.

This is what’s shown in the console, and I can confirm two /classifications POSTs in the network tab for the same subject.
queued

Questions:
1. I think this PR will prevent classifications from being lost, but should there also be a check for the current subject id in the failed classifications array to prevent duplication? A second classification from the same user in the scenario above would still be a unique classification I guess. Thoughts?
2. Does any of the info in ClassificationQueue's [Readme](https://github.com/zooniverse/front-end-monorepo/blob/master/packages/lib-classifier/src/store/utils/README.md) need to be updated?

**Comment by eatyourgreens** [comment]
Going back from Talk should reload the page, but that might not happen if you go back from the NextJS 404 page on the same domain.

**Comment by eatyourgreens** [comment]
I think going back from Talk loads a new subject. It’s hard to tell in mobile Safari. 

https://frontend.preview.zooniverse.org/projects/mschwamb/planet-hunters-ngts/classify/workflow/13694?demo=true

If it doesn’t, that’s a bug but a separate issue from the one here.

**Comment by eatyourgreens** [comment]
Subjects you’ve already seen should be handled here. I think that list is persisted across sessions but you may need to read the code to figure out how the classifier avoids showing you the same subject twice. 
https://github.com/zooniverse/front-end-monorepo/blob/master/packages/lib-classifier/src/helpers/subjectsSeenThisSession/subjectsSeenThisSession.js

**Comment by eatyourgreens** [comment]
The Readme for the classification queue is completely wrong about Chrome. We don’t use service workers at all. See https://github.com/zooniverse/front-end-monorepo/pull/510.

File: packages/lib-classifier/src/store/Classification/Classification.spec.js [hunk 1]
```diff
@@ -86,8 +86,9 @@ describe('Model > Classification', function () {
       snapshot = classification.toSnapshot()
     })
 
-    it('should not have an ID', function () {
```
[the remark this answers is not in the document]
Comment (goplayoutside3) [reply packages/lib-classifier/src/store/Classification/Classification.spec.js:89]: I see, thanks for the clarification!

**Review by goplayoutside3 — COMMENTED** [review]
Agreed that the same subject shouldn't be shown in cases like using the browser back button, and also agreed it's a discussion for a different PR. Good to know! If for some reason a user's classification fails and they classify on the same subject again, those would still be considered two unique classifications ✅ 

This PR is good to go once the queue's readme is updated!

**Comment by eatyourgreens** [comment]
The Readme hasn’t been touched since it was created in #213. I’m inclined to delete it, on favour of commenting the code. Is it useful? It obviously hasn’t been read in five years.

**Comment by eatyourgreens** [comment]
Unfortunately, this PR is still listed as 'review required'. Your review hasn't been submitted as a review.
Screenshot of the listing for this pull request, showing the status as 'review required'.

**Comment by eatyourgreens** [comment]
I've opened #4741 for the Readme, since none of the store utilities were documented.

Let's get this onto staging, so we can try it in an environment where the page unloads when you use 'Done & Talk'.

**Comment by eatyourgreens** [comment]
I've tried Done & Talk for a Gravity Spy subject in Firefox:
https://frontend.preview.zooniverse.org/projects/zooniverse/gravity-spy/classify/workflow/1610?demo=true

When I used the back button, the classifier refreshed with a new subject.

I haven't checked that in Chrome or Safari. Mobile Safari on iOS is definitely broken.

**Comment by eatyourgreens** [comment]
I've invoked admin privileges to merge this, so that I can run tests on staging.

**Comment by eatyourgreens** [comment]
No, there's no Talk set-up for a branch deploy.

**Pull request merged**